### PR TITLE
Add metric batch builder

### LINKED ISSUE #101: Add service.name as a first class metric attribute
`service.name` is a very useful, and almost-always-desired attribute for both metrics and spans. We should make it easy for users to set the `service.name` attribute on metric batches.

This could be a first-class field on the MetricBatchSender, the MetricBatch itself, or somehow on the metrics. A developer-friendly API should be the first consideration on figuring out the right way to implement this.

### PR DESCRIPTION
**Pull request by breedx-nr**
This adds a convenience builder to `MetricBatch` and `MetricBuffer` so that service name and instrumentation provider can be emphasized.  Hopefully this resolves #101 .

### CODE REVIEW

**Comment by jkwatson** [comment]
needs a rebase.
could you add an example that uses this mode of operation, as well (follow-up PR is fine)?

File: telemetry_examples/src/main/java/com/newrelic/telemetry/examples/TelemetryClientExample.java [hunk 1]
```diff
@@ -55,7 +55,12 @@ private static void sendSampleMetrics(
       TelemetryClient telemetryClient, Attributes commonAttributes) {
     long startTime = System.currentTimeMillis();
 
-    MetricBuffer metricBuffer = new MetricBuffer(commonAttributes);
+    MetricBuffer metricBuffer =
+        MetricBuffer.builder()
+            .serviceName("Sample Service")
+            .instrumentationProvider("Manual instrumentation")
+            .attributes(commonAttributes)
+            .build();
```
Comment (breedx-nr) [inline telemetry_examples/src/main/java/com/newrelic/telemetry/examples/TelemetryClientExample.java]: @jkwatson Here's a pretty uninspired example that I piggy backed on.

File: telemetry_core/src/main/java/com/newrelic/telemetry/metrics/CommonAttributesBuilder.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import com.newrelic.telemetry.Attributes;
+
+/** Encapsulates the building of commonly used attributes. */
+public class CommonAttributesBuilder {
+
+  private Attributes attributes = new Attributes();
+  private String serviceName;
+  private String instrumentationProvider;
+
+  public CommonAttributesBuilder attributes(Attributes attributes) {
```
Comment (jkwatson) [inline telemetry_core/src/main/java/com/newrelic/telemetry/metrics/CommonAttributesBuilder.java]: detailed javadoc is needed throughout this class. This is a public API and should be documented as such. *Especially* the instrumentation provider, the meaning of which will not be obvious to end users.

File: telemetry_core/src/main/java/com/newrelic/telemetry/metrics/MetricBatch.java [hunk 3]
```diff
@@ -23,4 +24,53 @@ public MetricBatch(Collection<Metric> metrics, Attributes commonAttributes) {
...
+     *
+     * @param serviceName - The name of the service
+     */
+    public Builder serviceName(String serviceName) {
+      commonAttributesBuilder.serviceName(serviceName);
+      return this;
+    }
+
+    /**
+     * Optional. Specify the name of the instrumentation that provides the metrics. The
```
Comment (jkwatson) [inline telemetry_core/src/main/java/com/newrelic/telemetry/metrics/MetricBatch.java]: I think examples will be useful in these docs.  Also, it's actually not the name of the instrumentation. It's the library that provides the instrumentation APIs.  For example, "micrometer" would be the instrumentation provider, which is a general instrumentation library, not a provider of specific instrumentation.

File: telemetry_core/src/main/java/com/newrelic/telemetry/metrics/MetricBatch.java [hunk 4]
```diff
@@ -23,4 +24,53 @@ public MetricBatch(Collection<Metric> metrics, Attributes commonAttributes) {
...
+     * instrumentation provider will be included in all common attributes.
+     *
+     * @param instrumentationProvider - The instrumentation provider name
+     */
+    public Builder instrumentationProvider(String instrumentationProvider) {
+      commonAttributesBuilder.instrumentationProvider(instrumentationProvider);
+      return this;
+    }
+
+    public Builder attributes(Attributes attributes) {
```
Comment (jkwatson) [inline telemetry_core/src/main/java/com/newrelic/telemetry/metrics/MetricBatch.java]: javadoc needed

File: telemetry_core/src/main/java/com/newrelic/telemetry/metrics/MetricBatch.java [hunk 5]
```diff
@@ -23,4 +24,53 @@ public MetricBatch(Collection<Metric> metrics, Attributes commonAttributes) {
...
+      commonAttributesBuilder.attributes(attributes);
+      return this;
+    }
+
+    public Builder metrics(Collection<Metric> metrics) {
```
Comment (jkwatson) [inline telemetry_core/src/main/java/com/newrelic/telemetry/metrics/MetricBatch.java]: javadoc needed

File: telemetry_core/src/main/java/com/newrelic/telemetry/metrics/MetricBatch.java [hunk 6]
```diff
@@ -23,4 +24,53 @@ public MetricBatch(Collection<Metric> metrics, Attributes commonAttributes) {
...
+      this.metrics = metrics;
+      return this;
+    }
+
+    public MetricBatch build() {
```
Comment (jkwatson) [inline telemetry_core/src/main/java/com/newrelic/telemetry/metrics/MetricBatch.java]: javadoc needed

File: telemetry_core/src/main/java/com/newrelic/telemetry/metrics/MetricBuffer.java [hunk 7]
```diff
@@ -75,4 +75,47 @@ public MetricBatch createBatch() {
...
+     * instrumentation provider will be included in all common attributes.
+     *
+     * @param instrumentationProvider - The instrumentation provider name
+     */
+    public Builder instrumentationProvider(String instrumentationProvider) {
+      commonAttributesBuilder.instrumentationProvider(instrumentationProvider);
+      return this;
+    }
+
+    public MetricBuffer build() {
```
Comment (jkwatson) [inline telemetry_core/src/main/java/com/newrelic/telemetry/metrics/MetricBuffer.java]: javdoc needed

File: telemetry_core/src/main/java/com/newrelic/telemetry/metrics/MetricBuffer.java [hunk 8]
```diff
@@ -75,4 +75,47 @@ public MetricBatch createBatch() {
...
+     *
+     * @param serviceName - The name of the service
+     */
+    public Builder serviceName(String serviceName) {
+      commonAttributesBuilder.serviceName(serviceName);
+      return this;
+    }
+
+    /**
+     * Optional. Specify the name of the instrumentation that provides the metrics. The
```
Comment (jkwatson) [inline telemetry_core/src/main/java/com/newrelic/telemetry/metrics/MetricBuffer.java]: see my comment on the other method for this.

**Comment by jkwatson** [comment]
Implementation looks solid. More complete docs needed before shipping.

File: telemetry_core/src/main/java/com/newrelic/telemetry/metrics/CommonAttributesBuilder.java (hunk 2 above)
Comment (breedx-nr) [reply telemetry_core/src/main/java/com/newrelic/telemetry/metrics/CommonAttributesBuilder.java]: Sounds good, will do.

File: telemetry_core/src/main/java/com/newrelic/telemetry/metrics/MetricBatch.java (hunk 3 above)
Comment (breedx-nr) [reply telemetry_core/src/main/java/com/newrelic/telemetry/metrics/MetricBatch.java]: Agreed.

File: telemetry_core/src/main/java/com/newrelic/telemetry/metrics/CommonAttributesBuilder.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+   * @param serviceName - The name of the service.
+   * @return this builder
+   */
+  public CommonAttributesBuilder serviceName(String serviceName) {
+    this.serviceName = serviceName;
+    return this;
+  }
+
+  /**
+   * Specifies the name of the provider of the instrumentation to this builder. Users doing manual
```
Comment (jkwatson) [inline telemetry_core/src/main/java/com/newrelic/telemetry/metrics/CommonAttributesBuilder.java]:
```suggestion
   * Specifies the name of the provider of the instrumentation to this builder, for example "micrometer" or "opentelemetry". Users doing manual
```

Comment (jkwatson) [reply telemetry_core/src/main/java/com/newrelic/telemetry/metrics/CommonAttributesBuilder.java]: or just copypasta from the other one that looks great. :)

**Comment by jkwatson** [comment]
just one last little tweak, then :shipit:

**Pull request merged**